qlib/Logger.qm

```diff
@@ -0,0 +1,2847 @@
+                if (last_match.appender != e.appender) {
+                    bool found = False;
+                    foreach string id in (keys pendingEvents) {
+                        if (pendingEvents{id}.appender == e.appender) {
+                            last_match.id = id;
+                            last_match.appender = e.appender;
+                            found = True;
+                            break;
+                        }
+                    }
+                    if (!found) {
+                        # look into processing
+                        foreach string id in (keys processingEvents) {
+                            if (processingEvents{id}.appender == e.appender) {
+                                last_match.id = id;
+                                found = True;
+                                break;
+                            }
+                        }
+                        if (!found) {
+                            # add new entry
+                            last_match.id = string(lastId.next());
+                        }
+                        last_match.appender = e.appender;
+                        pendingEvents{last_match.id} = {
+                            "id": last_match.id,  # to get reason of this value duplicating key see problem description above ThreadPool::submit(id, ...) call
+                            "appender": e.appender,
+                            "events": (),
+                        };
+                    }
+                }
+                push pendingEvents{last_match.id}.events, e.('type', 'params');
+            }
+
+            # remove finished
+            while (finishedEvents.size() > 0) {
+                string id = finishedEvents.get(-1);
+                remove processingEvents{id};
+            }
+            # now try push pending
+            foreach string id in (keys pendingEvents) {
+                # remove meanwhile finished if any
+                while (finishedEvents.size() > 0) {
+                    remove processingEvents{finishedEvents.get(-1)};
+                }
+                # is available a free worker thread ?
+                if (maxThreads >= 0 && processingEvents.size() >= maxThreads) {
+                    return;
+                }
+                if (exists processingEvents{id}) {
+                    # we cannot trigger new thread if is being processed
+                    continue;
+                }
+                hash e = remove pendingEvents{id};
+                processingEvents{id} = e;
+                # When passing "id" then no copy is passed to submit pool queue and worker may get wrong value
+                # so we must pass value in new variable which won't change meanwhile when the value is to be used.
+                # It is feature of submit implementation related to parameter passing.
+                threadPool.submit(sub() {call_function(\worker(), e.id, e.appender, e.events); } );
```

Comment:
can u please explain it to me? I still don't get this problem.

Reply:
This is just an explanation why won't work reliably apparently the same statement:
threadPool.submit(sub() {call_function(\worker(), id, e.appender, e.events); } );